test/strings/basic.jl

```diff
@@ -612,17 +621,18 @@ end
             @test prevind(strs[i], 15, 4) == 10
             @test prevind(strs[i], 15, 10) == 0
             @test prevind(strs[i], 15, 9) == 1
-            @test prevind(strs[i], 15, 10) == 0
             @test prevind(strs[i], 16) == 15
             @test prevind(strs[i], 16, 1) == 15
             @test prevind(strs[i], 16, 2) == 14
-            @test prevind(strs[i], 20) == 15
-            @test prevind(strs[i], 20, 1) == 15
-            @test prevind(strs[i], 20, 10) == 1
-            @test_throws ArgumentError prevind(strs[i], 20, 0)
-
-            @test nextind(strs[i], -1) == 1
-            @test nextind(strs[i], -1, 1) == 1
+            @test prevind(strs[i], 20) == 19
+            @test prevind(strs[i], 20, 1) == 19
+            @test prevind(strs[i], 20, 10) == 7
```

Comment:
This kind of behavior sounds weird to me. So we assume that all imaginary characters take one code unit, and once we reach actual characters we use them? I can't imagine a situation where it would be useful, and it could hide bugs...

Reply:
I originally wrote the assumption that they all take one code unit, but realized that for substrings it's very convenient to just pass-through to the backing string which means that the "imaginary" characters are whatever size the actual characters in the backing string are.

Reply:
So basically that means that the behavior of `prevind`/`nextind` is undefined when an out of bounds index is passed, which isn't very Julian IMHO. It would be more standard to throw an error when an out of bounds index is passed, and enable the undefined behavior via `@inbounds`, but as you said it's convenient in some cases, so...

Reply:
It's not undefined – it satisfies lots of nice, predictable properties:

* `prevind(s, i) < i < nextind(s, i)`
* `prevind(s, i, n) ≤ i ≤ nextind(s, i, n)` for all `n ≥ 0`
* `prevind(s, nextind(s, i, n), n) == thisind(s, i)` for all `n > 0`
* `prevind(s, nextind(s, i, n), m) == nextind(s, i, n-m)` whenever `n > m`
* `nextind(s, prevind(s, i, n), m) == prevind(s, i, n-m)` whenever `n > m`

There is a very consistent philosophy to all of these changes: only throw an error when you _have_ to. For character iteration, the error only occurs when you try to ask about the code point of a malformed character – handling a character or a string that has malformed data is fine. For indexing, computing with out-of-bounds indices is fine, the only thing that's a problem is asking for a character or a string that is out of bounds. As I said before, we do this _all_ the time with array indices and it's not a problem this is just another form index arithmetic.

Reply:
I said it's undefined because the number of code units which are added depend on the parent string (if any), so for example it will be different for a plain `String` and for a `SubString`. So in practice if you pass an out-of-bounds index, call `nextind`/`prevind` a given number of times, and get an in-bounds index, it's totally unpredictable what character you get.

Array indices are quite different, since in that case the step is always 1.

Reply:
Where did this out-of-bounds array index come from? If you start in bounds for a given string and then end back in bounds you always get the same answer. Yes, the out of bounds values in the middle may be different but why does that matter?

Reply:
Yes, normally you won't pass an out of bounds index. But if that happens because of a bug it would have been nice to be able to catch that by default.

Reply:
Making out of bound code units always size one won’t help with that. We already discussed the bit about why allowing computing out of bounds values is useful. I guess what hasn’t been shown is that these functions need to accept out of bounds indices as well as produce them.